runtime/parachains/src/disputes.rs

```diff
@@ -0,0 +1,1963 @@
+
+				prev_participants.iter_ones()
+					.map(|i| (ValidatorIndex(i as _), SpamSlotChange::Dec))
+					.collect()
+			}
+			(true, true) | (true, false) => {
+				// nothing to do. (true, false) is also impossible.
+				Vec::new()
+			}
+		};
+
+		// 2. Check for fresh FOR supermajority. Only if not already concluded.
+		let slash_against = if let (false, true) = pre_post_contains(DisputeStateFlags::FOR_SUPERMAJORITY) {
+			if self.state.concluded_at.is_none() {
+				self.state.concluded_at = Some(self.now.clone());
+			}
+
+			// provide AGAINST voters to slash.
+			self.state.validators_against.iter_ones()
+				.map(|i| ValidatorIndex(i as _))
+				.collect()
+		} else {
+			Vec::new()
+		};
+
+		// 3. Check for fresh AGAINST supermajority.
+		let slash_for = if let (false, true) = pre_post_contains(DisputeStateFlags::AGAINST_SUPERMAJORITY) {
+			if self.state.concluded_at.is_none() {
+				self.state.concluded_at = Some(self.now.clone());
+			}
+
+			// provide FOR voters to slash.
+			self.state.validators_for.iter_ones()
+				.map(|i| ValidatorIndex(i as _))
+				.collect()
+		} else {
+			Vec::new()
+		};
+
+		ImportSummary {
+			state: self.state,
+			spam_slot_changes,
+			slash_against,
+			slash_for,
+			new_participants: self.new_participants,
+			new_flags: post_flags - pre_flags,
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Called by the initalizer to initialize the disputes module.
+	pub(crate) fn initializer_initialize(now: T::BlockNumber) -> Weight {
+		let config = <configuration::Module<T>>::config();
+
+		let mut weight = 0;
+		for (session_index, candidate_hash, mut dispute) in <Disputes<T>>::iter() {
+			weight += T::DbWeight::get().reads_writes(1, 0);
+
+			if dispute.concluded_at.is_none()
+				&& dispute.start + config.dispute_conclusion_by_time_out_period < now
+			{
+				Self::deposit_event(Event::DisputeTimedOut(candidate_hash));
+
+				dispute.concluded_at = Some(now);
+				<Disputes<T>>::insert(session_index, candidate_hash, &dispute);
+
+				if <Included<T>>::contains_key(&session_index, &candidate_hash) {
+					// Local disputes don't count towards spam.
+
+					weight += T::DbWeight::get().reads_writes(1, 1);
+					continue;
+				}
+
+				// mildly punish all validators involved. they've failed to make
+				// data available to others, so this is most likely spam.
+				SpamSlots::<T>::mutate(session_index, |spam_slots| {
+					let spam_slots = match spam_slots {
+						Some(ref mut s) => s,
+						None => return,
+					};
+
+					// also reduce spam slots for all validators involved, if the dispute was unconfirmed.
+					// this does open us up to more spam, but only for validators who are willing
+					// to be punished more.
+					//
+					// it would be unexpected for any change here to occur when the dispute has not concluded
+					// in time, as a dispute guaranteed to have at least one honest participant should
+					// conclude quickly.
+					let participating = decrement_spam(spam_slots, &dispute);
```

Comment:
Shouldn't we also be slashing the validators who didn't participate at all?

Reply:
We could replace ImOnline by GRANDPA by asking that validators vote on who they saw enough GRANDPA votes from.

It's messy doing anything similar here because they could always wait until the winning side emerges.  If we do refund for bugs, and do a good job of identifying bugs, then slashing rewards make non-participation unattractive, and modifying your code around this stuff becomes costly.

Reply:
@andresilva not really. if 2 validators cook up a dispute on something that doesn't exist, we shouldn't slash everybody as a result